docs/go: Document respected attributes better

doronbehar opened the pull request:
<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://hydra.nixos.org/job/nixpkgs/trunk/manual/latest/download/1/nixpkgs/manual.html#chap-reviewing-contributions
-->

###### Motivation for this change

~~Often, both for `shadowfox` and for `pistol` as well (see [this](https://github.com/NixOS/nixpkgs/pull/85452#pullrequestreview-395915896)), I experience issues setting `-ldflags` for go builds. I debugged this a bit, and for shadowfox, the package in test, `go install` _was_ called with my `-ldflags` (set from `buildFlagsArray`), but something unclear was wrong with the quoting, and either `go install` failed or `link` (Go's internal linker seen called if you use `NIX_DEBUG = 3`) didn't use it from some reason.~~

~~I had easy success with this new option for `gotify-cli` and `shadowfox-updater` - no double or mixed quotes usage.~~

It's hard to get `ldflags` / `buildFlagsArray` right, this PR improves the docs on the matter.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

zowoq commented:
Having two ways of setting these is confusing. I don't see that this is needed. Also this would mean `buildGoModule` handles `ldflags` differently from `buildGoPackage`.

zowoq commented:
Perhaps we should remove `buildFlags` and only use `buildFlagsArray`?

doronbehar commented:
> Having two ways of setting these is confusing.

I think that having both `buildFlags` and `buildFlagsArray` is just as confusing. And I had no success using either, with `buildFlags` I felt even further from success.

> Also this would mean buildGoModule handles ldflags differently from buildGoPackage.

I couldn't find any reference of `ldflags` used in `buildGoPackage`. Please correct me if my observation is wrong.

> I don't see that this is needed

The reality shows that sometimes it's super hard to get `buildFlagsArray` right so that the correct ldflags are used. Maybe something else is wrong there - maybe it's wrong shell quoting or something, I really can't put my finger on it.

@zowoq let's put this to the test: Let's take `gotify-cli` as an example. Currently, `$(nix-build -A gotify-cli)/bin/gotify --version` reports:

```
Version:   unknown
Commit:    unknown
BuildDate: unknown
```

Tell me if there's something wrong with this patch:

```patch
diff --git i/pkgs/tools/misc/gotify-cli/default.nix w/pkgs/tools/misc/gotify-cli/default.nix
index 9e9f3c84782..1b7f0fc9d57 100644
--- i/pkgs/tools/misc/gotify-cli/default.nix
+++ w/pkgs/tools/misc/gotify-cli/default.nix
@@ -19,6 +19,10 @@ buildGoModule rec {
     mv $out/bin/cli $out/bin/gotify
   '';
 
+  buildFlagsArray = [
+    "-ldflags='-X main.Version=${version} -X main.Commit=${version}'"
+  ];
+
   meta = with lib; {
     license = licenses.mit;
     homepage = "https://github.com/gotify/cli";
```

Why does applying this patch doesn't make `$(nix-build -A gotify-cli)/bin/gotify --version` spit the right variables?? If you think it's possible to achieve, please tell me how! You can use `NIX_DEBUG=6` to see both the bash commands being run by the builder, and the commands `link` is running when these flags are set.

doronbehar commented:
BTW that inline patch for `gotify-cli` uses almost verbatim what's used for `gotify-server`:

https://github.com/NixOS/nixpkgs/blob/1435f5be15d7c294016888abcffd2bc1aa084916/pkgs/servers/gotify/default.nix#L51-L53

zowoq commented:
>  I couldn't find any reference of ldflags used in buildGoPackage. Please correct me if my observation is wrong.

`buildGoPackage` has `buildFlags` and `buildFlagsArray` so `buildGoModule` will have different handling.

zowoq commented:
Doesn't need the single quotes.

```diff
diff --git a/pkgs/tools/misc/gotify-cli/default.nix b/pkgs/tools/misc/gotify-cli/default.nix
index 9e9f3c84782..355bc34262f 100644
--- a/pkgs/tools/misc/gotify-cli/default.nix
+++ b/pkgs/tools/misc/gotify-cli/default.nix
@@ -19,6 +19,10 @@ buildGoModule rec {
     mv $out/bin/cli $out/bin/gotify
   '';
 
+  buildFlagsArray = [
+    "-ldflags=-X main.Version=${version} -X main.Commit=${version}"
+  ];
+
   meta = with lib; {
     license = licenses.mit;
     homepage = "https://github.com/gotify/cli";
```

doronbehar commented:
> buildGoPackage has buildFlags and buildFlagsArray so buildGoModule will have different handling.

I'm not sure I understand the concern / issue, but `buildGoModule` also uses both `buildFlags` and `buildFlagsArray`. Perhaps it'd be nice to do the same thing I done for `buildGoModule`, but for `buildGoPackage` as well.

zowoq commented:
This style works as well:

```diff
diff --git a/pkgs/tools/misc/gotify-cli/default.nix b/pkgs/tools/misc/gotify-cli/default.nix
index 9e9f3c84782..fde9dfb9f14 100644
--- a/pkgs/tools/misc/gotify-cli/default.nix
+++ b/pkgs/tools/misc/gotify-cli/default.nix
@@ -19,6 +19,12 @@ buildGoModule rec {
     mv $out/bin/cli $out/bin/gotify
   '';
 
+  buildFlagsArray = ''
+    -ldflags=
+    -X main.Version=${version}
+    -X main.Commit=${version}
+  '';
+
   meta = with lib; {
     license = licenses.mit;
     homepage = "https://github.com/gotify/cli";
```

doronbehar commented:
DAMN. I would have never guessed it by myself, that quotes for spaces are not needed. This should be documented. I'm turning over the PR to a docs PR.

zowoq commented on doc/languages-frameworks/go.xml:
```diff
@@ -152,7 +152,7 @@ deis = buildGoPackage rec {
     <callout arearefs='ex-buildGoPackage-5'>
      <para>
-      <varname>buildFlags</varname> is a list of flags passed to the go build command.
+       <varname>buildFlags</varname> is a list of flags passed to the go build command, explained with more examples <link xlink:href="#ssec-go-common-attributes">here</link>.
```
I'd remove this line and line 122 entirely and just have "Attributes used by the builders" section otherwise it still looks to be specific to `buildGoPackage`.

Later we could de-dupe `subPackages`, `deleteVendor` and add other common attrs (e.g. `excludedPackages`).

doronbehar replied:
I agree.

doronbehar replied:
I removed the commonly respected attributes from the examples to a common section where all of them are explained.

zowoq commented on doc/languages-frameworks/go.xml:
```diff
@@ -221,4 +221,34 @@ done
  </section>
+ <section xml:id="ssec-go-common-attributes">
+  <title>Attributes used by the builders</title>
+   <para>Both <link xlink:href="#ssec-go-modules"><varname>buildGoModule</varname></link> and <link xlink:href="#ssec-go-modules"><varname>buildGoPackage</varname></link> Read the attributes <varname>buildFlagsArray</varname> and <varname>buildFlags</varname> to set build flags supported by <varname>go</varname>'s commands such as <varname>go install</varname> and <varname>go build</varname>. Usu ...
```
Maybe remove `buildFlags` and just have `buildFlagsArray`?

doronbehar replied:
I think it'd be a lie to not mention `buildFlags` at all, unless we also remove it from the builder. However, we can recommend in the docs to use the array.

zowoq commented on doc/languages-frameworks/go.xml:
```diff
@@ -221,4 +179,73 @@ done
+    </term>
+    <listitem>
+     <para>
+      These attributes set build flags supported by <varname>go</varname>'s commands such as <varname>go install</varname> and <varname>go build</varname>. We recommend to use <varname>buildFlagsArray</varname>. The most common usecase of these attributes, is to make the resulting executable aware of it's own version, or alike. For example:
```
```suggestion
      These attributes set build flags supported by <varname>go build</varname>. We recommend using <varname>buildFlagsArray</varname>. The most common use case of these attributes is to make the resulting executable aware of its own version. For example:
```

zowoq commented on doc/languages-frameworks/go.xml:
```diff
@@ -221,4 +179,73 @@ done
+      These attributes set build flags supported by <varname>go</varname>'s commands such as <varname>go install</varname> and <varname>go build</varname>. We recommend to use <varname>buildFlagsArray</varname>. The most common usecase of these attributes, is to make the resulting executable aware of it's own version, or alike. For example:
+     </para>
+     <example xml:id='ex-goBuildFlags-nospaces'>
+      <title>buildFlagsArray snippet</title>
```
```suggestion
      <title>buildFlagsArray example</title>
```

zowoq commented on doc/languages-frameworks/go.xml:
```diff
@@ -221,4 +179,73 @@ done
+     <calloutlist>
+      <callout arearefs='ex-goBuildFlags-1'>
+       <para>
+        Note how no single quotes are needed!
```
```suggestion
        Note: single quotes are not needed.
```

zowoq commented on doc/languages-frameworks/go.xml:
```diff
@@ -221,4 +179,73 @@ done
+      This form should also work:
+     </para>
+     <example xml:id='ex-goBuildFlags-noarray'>
+      <title>buildFlagsArray snippet</title>
```
```suggestion
      <title>buildFlagsArray example</title>
```

zowoq commented on doc/languages-frameworks/go.xml:
```diff
@@ -221,4 +179,73 @@ done
+        -X main.Version=${version}
+        -X main.Commit=${version}
+      '';
+    </programlisting>
```
Indentation is off.

doronbehar replied:
Fixed.

zowoq commented on doc/languages-frameworks/go.xml:
```diff
@@ -221,4 +179,73 @@ done
+    </term>
+    <listitem>
+     <para>
+      Removes the pre-existing vendor directory and fetches the dependencies. This should only be used if the dependencies included in the vendor folder are broken or incomplete.
```
This no longer applies to either.

```suggestion
      Removes the pre-existing vendor directory. This should only be used if the dependencies included in the vendor folder are broken or incomplete.
```

zowoq commented on doc/languages-frameworks/go.xml:
```diff
@@ -221,4 +179,73 @@ done
+       </para>
+      </callout>
+     </calloutlist>
+     <para>
+      This form should also work:
+     </para>
```
```suggestion
```

doronbehar replied:
:+1:.

doronbehar committed:
docs/go: Add examples for and explain buildFlags

Move common attributes treated by both buildGoModule and buildGoPackage
to a separate section, out of the examples' "callouts".

Co-authored-by: zowoq <59103226+zowoq@users.noreply.github.com>

doronbehar commented:
Squashed all changes.

zowoq commented:
Thanks @doronbehar!
